### PR TITLE
Feature/pretty printer

### PR DESCRIPTION
**Pull request by solomon-b**
Implements a pretty printer and brings back the json roundtrip tests. In the process I discovered a bunch of subtle bugs related to unicode encoding and span construction.

### CODE REVIEW

File: src/Kriti/Eval.hs [hunk 1]
```diff
@@ -12,82 +13,102 @@ import qualified Data.Text.Encoding as TE
...
+typoOfJSON J.Object {} = "Object"
+typoOfJSON J.Array {} = "Array"
+typoOfJSON J.String {} = "String"
+typoOfJSON J.Number {} = "Number"
+typoOfJSON J.Bool {} = "Boolean"
+typoOfJSON J.Null = "Null"
 
 eval :: ValueExt -> ExceptT EvalError (Reader Ctxt) J.Value
-eval = \case
+eval term = case term of
```
Comment (jkachmar) [inline src/Kriti/Eval.hs:86]: Why the change from `\case` to an explicit argument? Nothing wrong with that, just curious.

File: src/Kriti/Eval.hs [hunk 2]
```diff
@@ -12,82 +13,102 @@ import qualified Data.Text.Encoding as TE
...
-serializeType J.Null = "Null"
+typoOfJSON :: J.Value -> T.Text
+typoOfJSON J.Object {} = "Object"
+typoOfJSON J.Array {} = "Array"
+typoOfJSON J.String {} = "String"
+typoOfJSON J.Number {} = "Number"
+typoOfJSON J.Bool {} = "Boolean"
+typoOfJSON J.Null = "Null"
 
 eval :: ValueExt -> ExceptT EvalError (Reader Ctxt) J.Value
```
Comment (jkachmar) [inline src/Kriti/Eval.hs:85]: I haven't gotten a chance to survey the codebase lately, but depending on how common this transformer is it might make sense to just pack it up into a `KritiM` newtype at some point.

I figure most interpreters/evaluators end up with some sort of environmental context (`ReaderT`) and error handler (`ExceptT`) anyway, so it seems like it would make sense.

File: src/Kriti/Parser/Token.hs [hunk 3]
```diff
@@ -151,5 +155,56 @@ instance Located ValueExt where
...
+          p (k, v) = pretty k <> colon <+> pretty v <> comma
+       in braces $ braces $ foldMap p $ Compat.toList km
+    Array _ vec -> pretty $ V.toList vec
+    String _ txt -> dquotes (pretty txt)
+    Number _ sci -> pretty $ show sci
+    Boolean _ b -> pretty b
+    Null _ -> "null"
+    StringTem _ vec ->
+      dquotes $
+        flip foldMap vec $ \case
```
Comment (jkachmar) [inline src/Kriti/Parser/Token.hs:174]: This isn't even a suggestion, more just letting you know of an alternative pattern:

With `-XBlockArguments` you can write `vec & foldMap \case [...]` which (IMO) reads a bit more nicely (almost like method call syntax).

File: src/Kriti/Eval.hs (hunk 1 above)
Comment (solomon-b) [reply src/Kriti/Eval.hs:86]: Ooops, i was passing in the original source bytestring explicitly and forgot to go back to `\case` when i moved the source into the Reader context.

File: src/Kriti/Eval.hs (hunk 2 above)
Comment (solomon-b) [reply src/Kriti/Eval.hs:85]: Yeah this transformer stack is extremely common for intepreters. i didn't bother with a newtype cause this interpreter is so small. We only have 2 functions with that type. I figured I would rethink this decision later if the interpreter starts to grow in complexity.

File: src/Kriti/Parser/Token.hs (hunk 3 above)
Comment (solomon-b) [reply src/Kriti/Parser/Token.hs:174]: oh I like that and actually have BlockArguments enabled. I always forget to use them
